### PR TITLE
Add FleeGoal

### PR DESCRIPTION
**Pull request by lionaneesh**
A goal which can be configured to run away to a possible safepoint if X mobs attack. Also, later I want to add checks based on HP/Mana which can be configured via the Json.

### CODE REVIEW

**Comment by Xian55** [comment]
Gooday, thanks for the contribution i will look into this sometime soon tomorrow. On the other hand i want to try out how it feels and probably going to help eliminate the warnings.

**Comment by Xian55** [comment]
There's a way to get notified about upon a mob is died by the player. There's a decoupled messaging system setup where `GoapGoals` be able to communicate with each other without knowing each other.

I will definitely going to improve the PR that way.

File: Core/GOAP/GoapAgentState.cs [hunk 1]
```diff
@@ -9,4 +11,5 @@ public sealed class GoapAgentState
     public int ConsumableCorpseCount { get; set; }
     public int LastCombatKillCount { get; set; }
     public bool Gathering { get; set; }
+    public LinkedList<Vector3> safeLocations = new LinkedList<Vector3>();
```
Comment (Xian55) [inline Core/GOAP/GoapAgentState.cs]: In my humble opinion, making `safeLocations` as a `LinkedList<Vector3>` seems like an overkill. 

In my humble opinion, a safe spot would be a location what the player was already stepped on in the **recent past**. 

While the player is outside of combat, periodically `X` seconds or `X` milliseconds, a location can be marked as safe.

Upon entering the `FleeGoal` the goal execution would be really simple, from a `Stack` collection just pops safe locations until leaves combat.

There should be some guard clause that the `safeLocations` should contain about `Y` seconds worth of time in order to make sure that do not run out of `safeLocations` this can be checked in the `CanRun()`. 

Basically `safeLocations.Count > 0`

Essentially while the `safeLocations` has content we do not need to reach for the `Navigation Component`.

It should be a good idea to find out, after entering combat with a mob, how many seconds does it take till the enemy exit combat.

**Comment by Xian55** [comment]
Hello, i've slightly refactored the code to be more simpler. It should give a better integration with the existing framework.

The biggest benefit probably is to see the path while in the `FleeGoal`. I left some TODOs in the code a few things needs to be figured out.

**Comment by Xian55** [comment]
Currently when the `SafeLocations` is running out of saved elements, the player just keep moving forward until leaves combat.

However this can be dangerous 👀

**Comment by lionaneesh** [comment]
Wow thanks a lot @Xian55 for your updates. Shall we merge this now or is there something we need to still figure out?

**Comment by Xian55** [comment]
I left some todo-s in the code these are issues which needs to be solved in order to the feature to be more resilient.

Feel fee to tackle them if you feel the urge, however couple of hours later i will try to spare some time and thinking about it!

**Comment by Xian55** [comment]
I've added a way to define custom rules when the `FleeGoal` is applicable. This way the Goal is only restricted by a simple rule by being in combat. The rest of the rules can be defined by the user.

I would preferably use the following condition

```json
"Flee": {
    "Sequence": [
        {
            "Name": "Flee",
            "Requirement": "MobCount > 1 && Health% < 50"
        }
    ]
},
```

I would start to running away from the combat while fighting with 2 mobs and the player health is below 50%.

Thanks for the contribution! 🙇
